src/util/metadata/header_metadata.rs

```diff
@@ -29,668 +27,1004 @@ pub struct HeaderMetadataSpec {
+    pub fn fetch_or<T: MetadataValue>(
+        &self,
+        object: ObjectReference,
+        val: T,
+        order: Ordering,
+    ) -> T {
+        #[cfg(debug_assertions)]
+        self.assert_spec::<T>();
+        if self.num_of_bits < 8 {
+            let (_, mask) = self.get_shift_and_mask_for_bits();
+            let new_val = val.to_u8().unwrap() & mask;
+            // We do not need to use fetch_ops_on_bits(), we can just set irrelavent bits to 1, and do fetch_and
+            let old_raw_byte =
+                unsafe { <u8 as MetadataValue>::fetch_or(self.meta_addr(object), new_val, order) };
+            let old_val = self.get_bits_from_u8(old_raw_byte);
+            FromPrimitive::from_u8(old_val).unwrap()
+        } else {
+            unsafe { T::fetch_or(self.meta_addr(object), val, order) }
+        }
+    }
 
-        // let new_metadata = ((val as u8) << bit_shift);
-        let byte_addr = object.to_address() + byte_offset;
-        loop {
+    /// This function provides a default implementation for the `fetch_update` method from the `ObjectModel` trait.
+    /// The semantics is the same as Rust's `fetch_update` on atomic types.
+    #[inline(always)]
+    pub fn fetch_update<T: MetadataValue>(
+        &self,
+        object: ObjectReference,
+        set_order: Ordering,
+        fetch_order: Ordering,
+        mut f: impl FnMut(T) -> Option<T> + Copy,
+    ) -> std::result::Result<T, T> {
+        #[cfg(debug_assertions)]
+        self.assert_spec::<T>();
+        if self.num_of_bits < 8 {
+            let byte_addr = self.meta_addr(object);
             unsafe {
-                let old_byte = byte_addr.atomic_load::<AtomicU8>(order);
-                let old_metadata = (old_byte & mask) >> bit_shift;
-                // new_metadata may contain overflow and should be and with the mask
-                let new_metadata = (old_metadata + val as u8) & (mask >> bit_shift);
-                let new_byte = (old_byte & !mask) | ((new_metadata as u8) << bit_shift);
-                if byte_addr
-                    .compare_exchange::<AtomicU8>(old_byte, new_byte, order, order)
-                    .is_ok()
-                {
-                    return old_metadata as usize;
-                }
+                <u8 as MetadataValue>::fetch_update(
+                    byte_addr,
+                    set_order,
+                    fetch_order,
+                    |raw_byte: u8| {
+                        let old_metadata = self.get_bits_from_u8(raw_byte);
+                        match f(FromPrimitive::from_u8(old_metadata).unwrap()) {
+                            Some(new_val) => {
+                                let new_metadata =
+                                    self.truncate_bits_in_u8(new_val.to_u8().unwrap());
+                                Some(self.set_bits_to_u8(raw_byte, new_metadata))
+                            }
+                            None => None,
+                        }
```

Comment:
This is the pattern for `Option::map`.

```suggestion
                        f(FromPrimitive::from_u8(old_metadata).unwrap()).map(|new_val| {
                                let new_metadata =
                                    self.truncate_bits_in_u8(new_val.to_u8().unwrap());
                                self.set_bits_to_u8(raw_byte, new_metadata)
                        }
```

Reply:
Done